File: src/main/java/net/minestom/server/extras/query/Query.java
package net.minestom.server.extras.query;

import io.netty.bootstrap.Bootstrap;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.*;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.DatagramPacket;
import io.netty.channel.socket.nio.NioDatagramChannel;
import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import net.minestom.server.MinecraftServer;
import net.minestom.server.event.EventDispatcher;
import net.minestom.server.extras.query.event.BasicQueryEvent;
import net.minestom.server.extras.query.event.FullQueryEvent;
import net.minestom.server.extras.query.response.BasicQueryResponse;
import net.minestom.server.extras.query.response.FullQueryResponse;
import net.minestom.server.network.NetworkBuffer;
import net.minestom.server.timer.Task;
import net.minestom.server.utils.time.TimeUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Random;

/**
 * GameSpy4 Query Protocol implementation backed by Netty UDP rather than
 * {@code java.net.DatagramSocket}.
 *
 * <p>No {@code java.nio.channels.*} or {@code sun.misc.Unsafe} references;
 * raw byte I/O is performed through Netty's {@link ByteBuf}.
 *
 * @see <a href="https://minecraft.wiki/w/Minecraft_Wiki:Projects/wiki.vg_merge/Query">
 *     Minecraft wiki – Query protocol</a>
 */
public class Query {

    public static final Charset CHARSET = StandardCharsets.ISO_8859_1;

    private static final Logger LOGGER   = LoggerFactory.getLogger(Query.class);
    private static final Random RANDOM   = new Random();

    /** challenge-token -> sender address */
    private static final Int2ObjectMap<SocketAddress> CHALLENGE_TOKENS =
            Int2ObjectMaps.synchronize(new Int2ObjectOpenHashMap<>());

    private static volatile boolean       started;
    private static volatile Channel       udpChannel;
    private static volatile EventLoopGroup eventLoopGroup;
    private static volatile Task          task;

    private Query() {}

    /**
     * Starts the query system on an OS-assigned port.
     *
     * @return the bound port
     * @throws IllegalArgumentException if already running
     */
    public static int start() {
        if (udpChannel != null) throw new IllegalArgumentException("System is already running");
        start(0);
        return ((InetSocketAddress) udpChannel.localAddress()).getPort();
    }

    /**
     * Starts the query system on the given {@code port}.
     *
     * @return {@code true} on success, {@code false} if already running or
     *         if the port could not be bound
     */
    public static boolean start(int port) {
        if (udpChannel != null) return false;

        final EventLoopGroup group = new NioEventLoopGroup(1);
        final Bootstrap bootstrap = new Bootstrap()
                .group(group)
                .channel(NioDatagramChannel.class)
                .handler(new QueryHandler());

        try {
            udpChannel      = bootstrap.bind(port).sync().channel();
            eventLoopGroup  = group;
            started         = true;
        } catch (Exception e) {
            LOGGER.warn("Could not open the query port!", e);
            group.shutdownGracefully();
            return false;
        }

        task = MinecraftServer.getSchedulerManager()
                .buildTask(CHALLENGE_TOKENS::clear)
                .repeat(30, TimeUnit.SECOND)
                .schedule();

        return true;
    }

    /**
     * Stops the query system.
     *
     * @return {@code true} if it was running, {@code false} otherwise
     */
    public static boolean stop() {
        if (!started) return false;

        started = false;
        if (udpChannel != null) {
            udpChannel.close().awaitUninterruptibly(); udpChannel = null;
        }

        if (eventLoopGroup != null) {
            eventLoopGroup.shutdownGracefully();        eventLoopGroup = null;
        }

        if (task != null) {
            task.cancel(); task = null;
        }

        CHALLENGE_TOKENS.clear();
        return true;
    }

    /** @return {@code true} if the query system is currently running */
    public static boolean isStarted() {
        return started;
    }

    /**
     * Handles inbound UDP datagrams and replies inline on the same Netty
     * event-loop thread — no extra thread needed.
     */
    @ChannelHandler.Sharable
    private static final class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket> {

        @Override
        protected void channelRead0(ChannelHandlerContext ctx, DatagramPacket msg) {
            final ByteBuf data   = msg.content();
            final InetSocketAddress sender = msg.sender();

            // Check magic 0xFEFD
            if (data.readableBytes() < 3) return;
            final int magic = data.readUnsignedShort();
            if (magic != 0xFEFD) return;

            final byte type = data.readByte();

            if (type == 9) { // handshake
                if (data.readableBytes() < 4) return;
                final int sessionID      = data.readInt();
                final int challengeToken = RANDOM.nextInt();
                CHALLENGE_TOKENS.put(challengeToken, sender);

                final byte[] responseBytes = NetworkBuffer.makeArray(buf -> {
                    buf.write(NetworkBuffer.BYTE, (byte) 9);
                    buf.write(NetworkBuffer.INT, sessionID);
                    buf.write(NetworkBuffer.STRING_TERMINATED, String.valueOf(challengeToken));
                });

                send(ctx, sender, responseBytes);

            } else if (type == 0) { // stat
                if (data.readableBytes() < 8) return;
                final int sessionID      = data.readInt();
                final int challengeToken = data.readInt();
                final int remaining      = data.readableBytes();

                if (!CHALLENGE_TOKENS.containsKey(challengeToken)
                        || !CHALLENGE_TOKENS.get(challengeToken).equals(sender)) return;

                if (remaining == 0) { // basic query
                    final BasicQueryEvent event = new BasicQueryEvent(sender, sessionID);
                    EventDispatcher.callCancellable(event, () ->
                            sendQueryResponse(ctx, BasicQueryResponse.SERIALIZER,
                                    event.getQueryResponse(), sessionID, sender));
                } else if (remaining == 5) { // full query
                    final FullQueryEvent event = new FullQueryEvent(sender, sessionID);
                    EventDispatcher.callCancellable(event, () ->
                            sendQueryResponse(ctx, FullQueryResponse.SERIALIZER,
                                    event.getQueryResponse(), sessionID, sender));
                }
            }
        }

        @Override
        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
            if (started) {
                LOGGER.error("Error in query handler", cause);
            }
        }
    }

    private static <T> void sendQueryResponse(ChannelHandlerContext ctx,
                                              NetworkBuffer.Type<T> type, T response,
                                              int sessionID, InetSocketAddress sender) {
        final byte[] payload = NetworkBuffer.makeArray(buf -> {
            buf.write(NetworkBuffer.BYTE, (byte) 0);
            buf.write(NetworkBuffer.INT, sessionID);
            buf.write(type, response);
        });
        send(ctx, sender, payload);
    }

    private static void send(ChannelHandlerContext ctx,
                             InetSocketAddress recipient, byte[] data) {
        final ByteBuf buf = ctx.alloc().buffer(data.length).writeBytes(data);
        ctx.writeAndFlush(new DatagramPacket(buf, recipient))
                .addListener(f -> {
                    if (!f.isSuccess() && started) {
                        LOGGER.error("Failed to send query response to {}", recipient, f.cause());
                    }
                });
    }
}

File: src/main/java/net/minestom/server/extras/query/Query.kt
package net.minestom.server.extras.query

import it.unimi.dsi.fastutil.ints.Int2ObjectMaps
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
import net.minestom.server.MinecraftServer.Companion.schedulerManager
import net.minestom.server.MinecraftServer.Companion.server
import net.minestom.server.utils.time.Cooldown.isReady
import net.minestom.server.utils.time.Cooldown.refreshLastUpdate
import net.minestom.server.MinecraftServer.Companion.exceptionManager
import net.minestom.server.utils.validate.Check.stateCondition
import net.minestom.server.MinecraftServer.Companion.biomeManager
import net.minestom.server.world.biomes.BiomeManager.addBiome
import net.minestom.server.world.biomes.Biome.Companion.builder
import net.minestom.server.world.biomes.Biome.Builder.name
import net.minestom.server.world.biomes.Biome.Builder.build
import net.minestom.server.MinecraftServer.Companion.connectionManager
import net.minestom.server.MinecraftServer.Companion.brandName
import net.minestom.server.MinecraftServer.Companion.extensionManager
import net.minestom.server.utils.binary.BinaryWriter.writeBytes
import net.minestom.server.utils.binary.BinaryWriter.writeNullTerminatedString
import net.minestom.server.utils.binary.BinaryWriter.buffer
import net.minestom.server.utils.binary.BinaryWriter.writeByte
import net.minestom.server.utils.binary.BinaryWriter.writeInt
import net.minestom.server.utils.binary.BinaryWriter.toByteArray
import net.minestom.server.utils.binary.Writeable.write
import net.minestom.server.utils.binary.BinaryReader.readBytes
import net.minestom.server.utils.binary.BinaryReader.buffer
import net.minestom.server.utils.binary.BinaryReader.readRemainingBytes
import net.minestom.server.utils.binary.BinaryReader.readVarInt
import net.minestom.server.utils.binary.BinaryReader.readSizedString
import net.minestom.server.utils.binary.BinaryReader.readBoolean
import net.minestom.server.utils.time.Cooldown
import net.minestom.server.extras.lan.OpenToLANConfig
import net.minestom.server.extras.lan.OpenToLAN
import net.minestom.server.MinecraftServer
import java.lang.Runnable
import net.minestom.server.event.server.ServerListPingEvent
import net.minestom.server.ping.ServerListPingType
import java.io.IOException
import java.security.KeyPairGenerator
import java.security.NoSuchAlgorithmException
import net.minestom.server.extras.mojangAuth.MojangCrypt
import java.security.PublicKey
import javax.crypto.SecretKey
import java.io.UnsupportedEncodingException
import java.security.MessageDigest
import java.security.PrivateKey
import javax.crypto.spec.SecretKeySpec
import javax.crypto.IllegalBlockSizeException
import javax.crypto.BadPaddingException
import javax.crypto.NoSuchPaddingException
import javax.crypto.spec.IvParameterSpec
import java.security.GeneralSecurityException
import java.lang.RuntimeException
import net.minestom.server.extras.optifine.OptifineSupport
import net.minestom.server.world.biomes.BiomeManager
import net.minestom.server.utils.NamespaceID
import net.minestom.server.utils.binary.Writeable
import net.minestom.server.event.trait.CancellableEvent
import net.minestom.server.extras.query.event.QueryEvent
import net.minestom.server.extras.query.response.FullQueryResponse
import net.minestom.server.extras.query.response.BasicQueryResponse
import net.minestom.server.extras.query.response.QueryKey
import net.minestom.server.utils.binary.BinaryWriter
import net.kyori.adventure.text.serializer.plain.PlainComponentSerializer
import java.lang.StringBuilder
import java.nio.charset.Charset
import java.lang.IllegalArgumentException
import net.minestom.server.extras.query.event.BasicQueryEvent
import net.minestom.server.extras.query.event.FullQueryEvent
import net.minestom.server.extras.velocity.VelocityProxy
import net.minestom.server.utils.binary.BinaryReader
import net.minestom.server.entity.PlayerSkin
import net.minestom.server.event.EventDispatcher
import net.minestom.server.timer.Task
import net.minestom.server.utils.time.TimeUnit
import org.slf4j.LoggerFactory
import java.net.*
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * Utility class to manage responses to the GameSpy4 Query Protocol.
 *
 * @see [wiki.vg](https://wiki.vg/Query)
 */
class Query private constructor() {
    /**
     * Stops the query system.
     *
     * @return `true` if the query system was stopped, `false` if it was not running
     */
    fun stop(): Boolean {
        return if (!isStarted) {
            false
        } else {
            isStarted = false
            thread = null
            socket!!.close()
            socket = null
            task!!.cancel()
            CHALLENGE_TOKENS.clear()
            true
        }
    }

    companion object {
        val CHARSET = StandardCharsets.ISO_8859_1
        private val LOGGER = LoggerFactory.getLogger(Query::class.java)
        private val RANDOM = Random()
        private val CHALLENGE_TOKENS = Int2ObjectMaps.synchronize(Int2ObjectOpenHashMap<SocketAddress>())

        /**
         * Checks if the query system has been started.
         *
         * @return `true` if it has been started, `false` otherwise
         */
        @Volatile
        var isStarted = false
            get() = Companion.field
            private set

        @Volatile
        private var socket: DatagramSocket? = null

        @Volatile
        private var thread: Thread? = null

        @Volatile
        private var task: Task? = null

        /**
         * Starts the query system, responding to queries on a random port, logging if it could not be started.
         *
         * @return the port
         * @throws IllegalArgumentException if the system was already running
         */
        fun start(): Int {
            return if (socket != null) {
                throw IllegalArgumentException("System is already running")
            } else {
                val port = 0
                start(port)
                port
            }
        }

        /**
         * Starts the query system, responding to queries on a given port, logging if it could not be started.
         *
         * @param port the port
         * @return `true` if the query system started successfully, `false` otherwise
         */
        fun start(port: Int): Boolean {
            return if (socket != null) {
                false
            } else {
                try {
                    socket = DatagramSocket(port)
                } catch (e: SocketException) {
                    LOGGER.warn("Could not open the query port!", e)
                    return false
                }
                thread = Thread { run() }
                thread!!.start()
                isStarted = true
                task = schedulerManager
                    .buildTask { CHALLENGE_TOKENS.clear() }
                    .repeat(30, TimeUnit.SECOND)
                    .schedule()
                true
            }
        }

        private fun run() {
            val buffer = ByteArray(16)
            while (isStarted) {
                val packet = DatagramPacket(buffer, buffer.size)

                // try and receive the packet
                try {
                    socket!!.receive(packet)
                } catch (e: IOException) {
                    if (!isStarted) {
                        LOGGER.error("An error occurred whilst receiving a query packet.", e)
                        continue
                    } else {
                        return
                    }
                }

                // get the contents
                val data = ByteBuffer.wrap(packet.data)

                // check the magic field
                if (data.short and 0xFFFF != 0xFEFD) {
                    continue
                }

                // now check the query type
                val type = data.get()
                if (type.toInt() == 9) { // handshake
                    val sessionID = data.int
                    val challengeToken = RANDOM.nextInt()
                    CHALLENGE_TOKENS[challengeToken] = packet.socketAddress

                    // send the response
                    val response = BinaryWriter(32)
                    response.writeByte(9.toByte())
                    response.writeInt(sessionID)
                    response.writeNullTerminatedString(challengeToken.toString(), CHARSET)
                    try {
                        val responseData = response.toByteArray()
                        socket!!.send(DatagramPacket(responseData, responseData.size, packet.socketAddress))
                    } catch (e: IOException) {
                        if (!isStarted) {
                            LOGGER.error("An error occurred whilst sending a query handshake packet.", e)
                        } else {
                            return
                        }
                    }
                } else if (type.toInt() == 0) { // stat
                    val sessionID = data.int
                    val challengeToken = data.int
                    val sender = packet.socketAddress
                    if (CHALLENGE_TOKENS.containsKey(challengeToken) && CHALLENGE_TOKENS[challengeToken] == sender) {
                        val remaining = data.remaining()
                        if (remaining == 0) { // basic
                            val event = BasicQueryEvent(sender, sessionID)
                            EventDispatcher.callCancellable(event) {
                                sendResponse(
                                    event.queryResponse,
                                    sessionID,
                                    sender
                                )
                            }
                        } else if (remaining == 5) { // full
                            val event = FullQueryEvent(sender, sessionID)
                            EventDispatcher.callCancellable(event) {
                                sendResponse(
                                    event.queryResponse,
                                    sessionID,
                                    sender
                                )
                            }
                        }
                    }
                }
            }
        }

        private fun sendResponse(queryResponse: Writeable, sessionID: Int, sender: SocketAddress) {
            // header
            val response = BinaryWriter()
            response.writeByte(0.toByte())
            response.writeInt(sessionID)

            // payload
            queryResponse.write(response)

            // send!
            val responseData = response.toByteArray()
            try {
                socket!!.send(DatagramPacket(responseData, responseData.size, sender))
            } catch (e: IOException) {
                if (!isStarted) {
                    LOGGER.error("An error occurred whilst sending a query handshake packet.", e)
                }
            }
        }
    }
}